### ISSUE TITLE
Possible dereference of invalid pointer

### ISSUE THREAD

**Issue by mmuetzel** [gnu-octave/pkg-json#2]
It looks like 1.4.0 includes the same possible dereference of an invalid pointer that was included in the commit to core Octave.

See https://hg.savannah.gnu.org/hgweb/octave/rev/c7c69808356f for a possible fix.

**Comment by siko1056** [comment]
Thanks for the hint.  Did you receive a compiler error or warning with the original version?

**Comment by mmuetzel** [comment]
A build with clang on Ubuntu crashed while executing the test suite. See also the CI runners.
Since the code was dereferencing a pointer to an object that went out of scope, all compilers are probably affected. But crashes might occur randomly...

**Comment by siko1056** [comment]
I see, just for curiosity and for me to learn.

The pointer is pointing to an object created in the "root" function of function calls and a copy of that pointer is passed to all sub functions.  How can the pointer run out of scope?

**Comment by mmuetzel** [comment]
Basically, curly braces `{}` create a new scope in C/C++. Objects created inside that scope get "out of scope" after the execution leaves that scope. At that point, the corresponding memory is free to be overwritten. (That might or might not happen sooner or later. Hence the randomness.)
That is what was happening to `options_obj`.

**Comment by siko1056** [comment]
Thanks for the explanation.  My assertion is, that the object is living in the scope of the "root" Fjsondecode() and this scope is not erased until the very end of decoding :thinking: 

In other words is ptr in `b()` able to run out of scope?

```cc
void b (int* ptr)
{
  std::cout << *ptr << std::endl;
}

void main ()
{
  int val = 42;
  int* ptr = &val;
  b (ptr);
}
```

**Comment by mmuetzel** [comment]
That example is fine.
The implementation here is a little bit different:
``` c++
void main ()
{
  int* ptr = nullptr;
  if (true)
  {
    int val = 42;
    ptr = &val;
  }
  return;
}
```

After leaving the `if` block, `val` is out of scope. `ptr` points to memory that might be overwritten. Dereferencing it leads to undefined behavior and might crash the program.

Edit: If you'd like to continue allocating val on the stack, you could also fix it like this:
``` c++
void main ()
{
  int* ptr = nullptr;
  int val;
  if (true)
  {
    val = 42;
    ptr = &val;
  }
  return;
}
```

**Comment by siko1056** [comment]
Thank you @mmuetzel, I didn't see the forstest for the trees :sweat_smile: 

Your change is perfect, many thanks again and I'll push it here too.  Having no compiler warning or SegFault myself makes me greedy to understand the issue. Now I added this pattern to my list of things to take greater care about :wink:

**Comment by siko1056** [comment]
Closed by a68b7a463186c6056a8350de09bb3977c222e031.
